src/plugins/vis_builder/public/visualizations/vega/vega_lite_spec_builder.ts

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { buildVegaLiteEncoding } from './components/encoding';
+import { buildMarkForVegaLite, VegaMarkType } from './components/mark';
+import { buildTooltip } from './components/tooltip';
+import { buildLegend } from './components/legend';
+import { StyleState } from '../../application/utils/state_management';
+import { VegaLiteSpec, AxisFormats } from './utils/types';
+import { mapChartTypeToVegaType } from './utils/helpers';
+
+/**
+ * Builds a Vega-Lite specification based on the provided data, visual configuration, and style.
+ *
+ * @param {any} data - The data configuration, normally including axis formats and transformed data.
+ * @param {any} visConfig - The visual configuration including dimensions and display options.
+ * @param {StyleState} style - The StyleState defined in style slice.
+ * @returns {VegaLiteSpec} The complete Vega-Lite specification.
+ */
+export const generateVegaLiteSpec = (
+  data: any,
+  visConfig: any,
+  style: StyleState
+): VegaLiteSpec => {
+  const { dimensions, addLegend, legendPosition, addTooltip } = visConfig;
+  const { type } = style;
+  const vegaType = mapChartTypeToVegaType(type) as VegaMarkType;
+  const {
+    xAxisFormat,
+    xAxisLabel,
+    yAxisFormat,
+    yAxisLabel,
+    zAxisFormat,
+    series: transformedData,
+  } = data;
+
+  const formats: AxisFormats = {
+    xAxisFormat,
+    xAxisLabel,
+    yAxisFormat,
+    yAxisLabel,
+    zAxisFormat,
+  };
+
+  // Build the base Vega-Lite specification
+  const baseSpec: VegaLiteSpec = {
+    $schema: 'https://vega.github.io/schema/vega-lite/v5.json',
+    data: { values: transformedData },
+    mark: buildMarkForVegaLite(vegaType),
+    encoding: buildVegaLiteEncoding(dimensions, formats),
```

Comment:
nit: We are calling `buildVegaLiteEncoding(dimensions, formats)` three times separately in the function , we can store the value in first call and use it across places.

Reply:
Will polish in a follow up PR.